devicedefender/source/DeviceDefender.cpp

```diff
@@ -114,6 +190,189 @@ namespace Aws
             this->cancellationUserdata = nullptr;
         }
 
+        void ReportTask::RegisterCustomMetricNumber(
+            const Crt::String &&metricName,
+            CustomMetricNumberFunction &&metricFunc) noexcept
+        {
+            std::shared_ptr<CustomMetricNumber> data =
+                std::shared_ptr<CustomMetricNumber>(new CustomMetricNumber(metricFunc, m_allocator));
+            storedCustomMetrics.push_back(data);
+            aws_byte_cursor cursor = aws_byte_cursor_from_c_str(metricName.c_str());
+            aws_iotdevice_defender_config_register_number_metric(
+                m_taskConfig, &cursor, CustomMetricNumber::GetMetricFunction, data.get());
+        }
+
+        void ReportTask::RegisterCustomMetricNumberList(
+            const Crt::String &&metricName,
+            CustomMetricNumberListFunction &&metricFunc) noexcept
+        {
+            std::shared_ptr<CustomMetricNumberList> data =
+                std::shared_ptr<CustomMetricNumberList>(new CustomMetricNumberList(metricFunc, m_allocator));
+            storedCustomMetrics.push_back(data);
+            aws_byte_cursor cursor = aws_byte_cursor_from_c_str(metricName.c_str());
+            aws_iotdevice_defender_config_register_number_list_metric(
+                m_taskConfig, &cursor, CustomMetricNumberList::GetMetricFunction, data.get());
+        }
+
+        void ReportTask::RegisterCustomMetricStringList(
+            const Crt::String &&metricName,
+            CustomMetricStringListFunction &&metricFunc) noexcept
+        {
+            std::shared_ptr<CustomMetricStringList> data =
+                std::shared_ptr<CustomMetricStringList>(new CustomMetricStringList(metricFunc, m_allocator));
```

Comment:
when making shared_ptrs from CRT allocators, you'll want to use Aws::Crt::MakeShared